pkg/plugin/conntrack/_cprog/conntrack.c

```diff
@@ -0,0 +1,245 @@
+    value->lifetime = now + CT_CONNECTION_LIFETIME_NONTCP;
+    __u32 last_report = value->last_report;
+    // We will only report this packet if the report interval has passed.
+    if (now - last_report >= CT_REPORT_INTERVAL) {
+        value->last_report = now;
+        return true;
+    }
+    return false;
+}
+
+/**
+    * Process a packet and update the connection tracking map.
+    * @arg key The key to be used to lookup the connection in the map.
+    * @arg flags The flags of the packet.
+    * Returns true if the packet should be report to userspace. False otherwise.
+**/
+static __always_inline __attribute__((unused)) bool ct_process_packet(struct ct_v4_key key, __u8 flags) {    
+    // Lookup the connection in the map.
+    struct ct_value *value = bpf_map_lookup_elem(&retina_conntrack_map, &key);
+
+    // If the connection is not found based on given packet, there are a few possibilities:
+    // 1. The connection is new. This connection is either originated from the endpoint or destined to the endpoint.
+    // 2. The packet belong to an existing connection but in the reverse direction.
+    if (!value) { // The connection is not found in the forward direction. Check the reverse direction.
+        // Create a new key for the reverse direction.
+        struct ct_v4_key reverse_key;
+        __builtin_memset(&reverse_key, 0, sizeof(struct ct_v4_key));
+        reverse_key.src_ip = key.dst_ip;
+        reverse_key.dst_ip = key.src_ip;
+        reverse_key.src_port = key.dst_port;
+        reverse_key.dst_port = key.src_port;
+        reverse_key.proto = key.proto;
+        // Lookup the connection in the map based on the reverse key.
+        value = bpf_map_lookup_elem(&retina_conntrack_map, &reverse_key);
+        // If the connection is still not found, the connection is new.
+        if (!value) {
+            // Check what kind of protocol the packet is.
+            switch(key.proto) {
+                case IPPROTO_TCP: {
+                    // Check if the packet is a SYN packet.
+                    if (flags & TCP_SYN) {
+                        // Create a new connection.
+                        struct ct_value new_value;
+                        __builtin_memset(&new_value, 0, sizeof(struct ct_value));
+                        // Set the lifetime of the connection. Since this is a new connection, we will set the lifetime to SYN_TIMEOUT.
+                        __u64 now = bpf_mono_now();
+                        new_value.lifetime = now + CT_SYN_TIMEOUT;
+                        new_value.flags_seen = flags;
+                        new_value.last_report = now;
+                        new_value.is_closing = 0;
+                        bpf_map_update_elem(&retina_conntrack_map, &key, &new_value, BPF_ANY);
+                        return true;
+                    } else {
+                        // The packet is not a SYN packet and the connection corresponding to this packet is not found.
```

Comment:
we need to solve this otherwise we are missing observing an existing long running connections.